### PR TITLE
Preliminary draft for the profiler

### PR DESCRIPTION
**Pull request by mathrack**
After loading the GNU compiler (11.2.0) and the OpenMPI library (4.1.0), Caliper (2.8.0) was compiled with :
```
mkdir build
cd build
cmake -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_INSTALL_PREFIX=xxxxxxxx/caliper/caliper_2.8.0 -DWITH_FORTRAN=yes -DWITH_MPI=yes -DBUILD_TESTING=yes ../
make -j
make test
make install
```
One test is failing but the draft seems to work.

### CODE REVIEW

**Comment by mathrack** [comment]
```
mpirun -n 4 ./test2d 
 In auto-tuning mode......
 factors:            1           2           4
 p_row x p_col           2           2
  
 test2d completed
  
Path           Min time/rank Max time/rank Avg time/rank Time %    
decomp_2d_fin       0.000050      0.000052      0.000051  2.873644 
transp_y_x_r        0.000009      0.000010      0.000010  0.549373 
transp_z_y_r        0.000010      0.000015      0.000012  0.690238 
transp_y_z_r        0.000016      0.000023      0.000020  1.112833 
transp_x_y_r        0.000023      0.000143      0.000057  3.225806 
decomp_2d_init      0.000872      0.001010      0.000917 51.697422
```

```
mpirun -n 4 ./fft_test_r2c
 In auto-tuning mode......
 factors:            1           2           4
 p_row x p_col           2           2
  
 ***** Using the generic FFT engine *****
  
  
 ***** Using the generic FFT engine *****
  
  
  fft_test_r2c completed
  
Path           Min time/rank Max time/rank Avg time/rank Time %    
decomp_2d_fin       0.000030      0.000033      0.000031  1.735298 
fft_fin             0.000009      0.000011      0.000010  0.550888 
fft_c2r             0.000028      0.000034      0.000031  1.693982 
  transp_y_z_c      0.000007      0.000018      0.000012  0.661066 
  transp_x_y_c      0.000009      0.000017      0.000012  0.688610 
  transp_y_x_c      0.000007      0.000018      0.000012  0.633522 
  transp_z_y_c      0.000008      0.000017      0.000011  0.592205 
fft_r2c             0.000030      0.000038      0.000034  1.873020 
  transp_y_x_c      0.000007      0.000078      0.000043  2.368820 
  transp_z_y_c      0.000008      0.000082      0.000028  1.570032 
  transp_y_z_c      0.000011      0.000198      0.000104  5.743011 
  transp_x_y_c      0.000014      0.000198      0.000062  3.401735 
fft_init            0.000020      0.000139      0.000050  2.740669 
decomp_2d_init      0.000717      0.000885      0.000784 43.189643

```

File: Makefile [hunk 1]
```diff
@@ -8,12 +8,13 @@
...
 GIT_VERSION := $(shell git describe --tag --long --always)
 
-DEFS = -DDOUBLE_PREC -DVERSION=\"$(GIT_VERSION)\"
+DEFS = -DDOUBLE_PREC -DVERSION=\"$(GIT_VERSION)\" -DPROFILER
 
 LCL = local# local,lad,sdu,archer
 CMP = gcc# intel,gcc,nagfor,cray,nvhpc
 FFT = generic# fftw3,fftw3_f03,generic,mkl
 PARAMOD = mpi # multicore,gpu
+PROFILER = caliper# none, caliper (add -DPROFILER to DEFS if not none)
```
Comment (pbartholomew08) [inline Makefile:17]: Default to none

Comment (mathrack) [reply Makefile]: Fixed !

File: src/decomp_2d.f90 [hunk 2]
```diff
@@ -208,12 +208,31 @@ module decomp_2d
...
+  logical, save, public :: decomp_profiler_d2d = .true.
+
   ! public user routines
   public :: decomp_2d_init, decomp_2d_finalize, &
        transpose_x_to_y, transpose_y_to_z, &
        transpose_z_to_y, transpose_y_to_x, &
        decomp_info_init, decomp_info_finalize, partition, &
-       decomp_info_print, &
+       decomp_info_print, decomp_profiler_prep, &
+       decomp_profiler_start, decomp_profiler_end, &
```
Comment (pbartholomew08) [inline src/decomp_2d.f90:235]: Do these need to be #ifdef'd out? Are they defined if `PROFILER=none`?

Comment (mathrack) [reply src/decomp_2d.f90:235]: The interface is defined yes

**Review by pbartholomew08 — APPROVED** [review]

**Pull request merged**